src/core/async_cache.rs

```diff
@@ -42,37 +79,92 @@ impl<Key: Eq + Hash + Send + Clone, Value: Debug + Clone + Send, Error: Debug +
         key: Key,
         or_else: impl FnOnce() -> Pin<Box<dyn Future<Output = Result<Value, Error>> + 'a + Send>> + Send,
     ) -> Arc<Result<Value, Error>> {
-        if let Some(cache_value) = self.get_cache_value(&key) {
-            match cache_value {
-                CacheValue::Pending(tx) => tx.subscribe().recv().await.unwrap(),
-                CacheValue::Ready(value) => value,
+        if Behavior::should_cache() {
+            if let Some(cache_value) = self.get_cache_value(&key) {
+                match cache_value {
+                    CacheValue::Pending(tx) => tx.subscribe().recv().await.unwrap(),
+                    CacheValue::Ready(value) => value,
+                }
+            } else {
+                let (tx, rx) = broadcast::channel(1000);
+                self.cache
+                    .write()
+                    .unwrap()
+                    .insert(key.clone(), CacheValue::Pending(tx.clone()));
+                let result = Arc::new(or_else().await);
+                {
+                    let mut guard = self.cache.write().unwrap();
+                    if let Some(cache_value) = guard.get_mut(&key) {
+                        *cache_value = CacheValue::Ready(result.clone())
+                    }
+                }
+                tx.send(result.clone()).ok();
+                result
+            }
+        } else {
+            self.load_without_cache(key, or_else).await
+        }
+    }
+
+    async fn load_without_cache<'a>(
+        &self,
+        key: Key,
+        func: impl FnOnce() -> Pin<Box<dyn Future<Output = Result<Value, Error>> + 'a + Send>> + Send,
+    ) -> Arc<Result<Value, Error>> {
+        let mut subscriber = {
+            let lock = self.cache.read().unwrap();
+            if let Some(cache_value) = lock.get(&key) {
+                match cache_value {
+                    CacheValue::Pending(tx) => Some(tx.subscribe()),
+                    CacheValue::Ready(value) => unimplemented!(),
+                }
+            } else {
+                None
             }
+        };
+
+        if let Some(mut subscriber) = subscriber {
+            subscriber.recv().await.unwrap()
```

Comment:
I changed this part of the code so lock is hold now until we get the result of `tx.subscribe()`.

I believe in previous version of the code when we've been doing cloning of the `tx` there was small period of time when lock is released, but subscriber is not yet created. And if meantime the value is resolved and send to the subscribers then this current subscriber that will be created will never get any value at all leading to deadlock